1_fetch/src/sbtools_utils.R

```diff
@@ -0,0 +1,113 @@
+#'   the file on SB at the time we fetched it into this repo)
+read_sb_status <- function(status_file) {
+  read_csv(status_file, col_types = cols(filename=col_character(), .default=col_datetime()))
+}
+
+#' @param wait_interval time to wait before querying SB again. a difftime object, e.g., `as.difftime(1, units='days')`
+#' @param update_files vector of files to update, or missing to update all
+#' @param ignore_files vector of files to omit from the status tibble
+update_sb_status <- function(sb_id, status_file, wait_interval, update_files, ignore_files=c()) {
+  # get some starting status info into memory, reading in the previous status
+  # information if available
+  status <- if(file.exists(status_file)) {
+    read_sb_status(status_file) %>%
+      filter(!filename %in% ignore_files)
+  } else {
+    if(!dir.exists(dirname(status_file))) dir.create(dirname(status_file), recursive=TRUE)
+    tibble(filename='', date_uploaded=Sys.time(), date_checked=Sys.time(), date_fetched=Sys.time(), fetch_datestamp=Sys.time())[c(),]
+  }
+
+  # decide whether we have to do the update
+  has_update_files <- !missing(update_files)
+  prev_date_checked <- status %>% # find the earliest check date of those files
+    {if(has_update_files) { filter(., filename %in% update_files) } else .} %>%
+    pull(date_checked) %>%
+    {suppressWarnings(min(., na.rm = TRUE))}
+  if(nrow(status) == 0 || (has_update_files && length(update_files) == 0) || (Sys.time() - prev_date_checked) > wait_interval) {
+
+    # we need it, so query SB for status of all files (couldn't request info file-by-file without adding API calls)
+    message('Updating SB file status')
+    sb_secret_login()
+    new_status <- as.sbitem(sb_id)$files %>%
+      purrr::map_df(function(file_info) tibble::as_tibble(file_info[c('name','dateUploaded')])) %>%
+      rename(filename = name, date_uploaded = dateUploaded) %>%
+      filter(!filename %in% ignore_files) %>%
+      mutate(date_uploaded = as.POSIXct(date_uploaded, format='%Y-%m-%dT%H:%M:%SZ'), date_checked = Sys.time())
+
+    # adjust the status information based on the new query results
+    if(missing(update_files)) {
+      update_files <- new_status$filename
+    }
+    updates <- status %>%
+      select(filename, fetch_datestamp) %>%
+      full_join(new_status, by='filename') %>%
+      filter(filename %in% update_files)
+    out_status <- status %>%
+      filter(!(filename %in% update_files)) %>%
+      bind_rows(updates) %>%
+      arrange(filename)
+
+    # write the revised file
+    write_csv(out_status, status_file)
+
+  } else {
+    message('Using old SB file status')
+  }
+
+  return(status_file)
+}
+
+get_sb_outdated <- function(status_file) {
+  outdated <- read_sb_status(status_file) %>%
+    # I tried but can't find a hash field in the SB file info, so we'll use dateUploaded to determine currentness
+    mutate(outdated = (is.na(fetch_datestamp) | date_uploaded != fetch_datestamp)) %>%
```

Comment:
I'm likely misunderstanding the precise values of `date_uploaded` and `fetch_datestamp`, but I'm confused why you're only checking that they aren't equal here, rather than simply checking whether the `date_uploaded` is more recent than the `fetch_datestamp`?

Reply:
Huh. I agree it would probably be safe to only check for when `date_uploaded > fetch_datestamp`, but checking for `date_uploaded != fetch_datestamp` catches that situation and also the (much less likely; not even sure it's possible) situation where `date_uploaded < fetch_datestamp`. So I think what I currently have is also correct...chewing on whether switching to `>` would make the code clearer. 

The values of these variables are a bit weird - I'm taking them both from what ScienceBase reports about the files, though I request that information at different times, such that `date_uploaded == fetch_datestamp` will be a pretty common situation. `date_uploaded` is the date/time when the file was uploaded to ScienceBase, requested whenever we last updated `sb_status_csv` (probably within the last day). `fetch_datestamp` is that same information from ScienceBase, requested at the time we downloaded the file from ScienceBase into this repo. I struggled with variable naming - if there are alternative names that you think would be more clear, please let me know.

Reply:
ohhh I see. I misunderstood `fetch_datestamp` - I thought it was the datestamp of when the file had been 'fetched', e.g., downloaded, from ScienceBase. Trying to think what I would find more intuitive as a variable name. Maybe `check_datestamp`?

Reply:
Yeah, but both datestamps come from checks to SB...maybe `fetch_datestamp` -> `datestamp_on_fetch`, `date_uploaded` -> `datestamp_on_SB`? But it's still not super clear where the datestamp comes from in `datestamp_on_fetch` or when the datestamp is current as of in `datestamp_on_SB` - just hard to cram all that info into a variable name.

Reply:
That is tricky. I think `datestamp_on_fetch` and `datestamp_on_SB` are pretty good.